rgw: h3 frontend based on quiche library

cbodley opened the pull request:
https://github.com/cloudflare/quiche is a rust library with [C bindings](https://github.com/cloudflare/quiche/blob/master/quiche/include/quiche.h). this PR adds quiche as a cmake ExternalProject, using `cargo` to build it as a static library

like several other quic libraries, quiche depends on boringssl so this pr adds a librgw_h3_quiche.so and dlopens it to isolate the boringssl/quiche linkage from the rest of rgw/ceph

libcurl's cmake also supports a `USE_QUICHE` option, which should prove useful as an http/3 client for testing. the latest curl 7.85 doesn't build against any quiche release, but curl 7.84 does. so curl 7.84 and the latest quiche 0.14.0 were the selected versions

i've sketched out some of the rgw frontend based on quiche, boost::asio, and c++20 coroutines. by running each quic connection on a separate strand executor, i hope to avoid the need for explicit locking

TODO:
- [x] split quiche dependency into a shared library and dlopen it to isolate the boringssl linkage
- [ ] optimize i/o path for bandwidth
- [ ] optimize headers with custom allocator or replacement for `beast::http::fields`
- [ ] unit tests driven by a quiche client to cover interesting error paths
- [ ] functional tests driven by a curl client to exercise several streams over a single connection

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

cbodley committed:
rgw: add h3 frontend with quiche

Signed-off-by: Casey Bodley <cbodley@redhat.com>

yuvalif commented:
> https://github.com/cloudflare/quiche is a rust library with [C bindings](https://github.com/cloudflare/quiche/blob/master/quiche/include/quiche.h). this PR adds quiche as a cmake ExternalProject, using `cargo` to build it as a static library
> 

what were the reasons to prefer quiche over lsquic ?

> like several other quic libraries, quiche depends on boringssl so this pr builds on #48160 to enable it in ceph. as reported there, radosgw crashes on startup because the system libcurl is still linked to the system openssl library. to work around this, we build a custom libcurl against the same boringssl library
> 
> libcurl's cmake also supports a `USE_QUICHE` option, which should prove useful as an http/3 client for testing. the latest curl 7.85 doesn't build against any quiche release, but curl 7.84 does. so curl 7.84 and the latest quiche 0.14.0 were the selected versions
> 
> i've sketched out some of the rgw frontend based on quiche, boost::asio, and c++20 coroutines. by running each quic connection on a separate strand executor, i hope to avoid the need for explicit locking
> 

with the rest of the code we are using the spawn library with the optiobal_yield. i thought that the plan would be to refactor the frontends with c++20 coroutines in a later stage?

> Show available Jenkins commands

cbodley commented:
> what were the reasons to prefer quiche over lsquic ?

for me, the main differentiator was related to locking. with lsquic, the main API object is [lsquic_engine](https://lsquic.readthedocs.io/en/stable/apiref.html#c.lsquic_engine_t) which manages the state of all of its connections. every call into lsquic essentially requires a global lock on that engine object

in quiche, [Connection](https://docs.quic.tech/quiche/struct.Connection.html) is the top-level API object. there are 'global' `quiche_config` and `quiche_h3_config` objects but they're immutable. because each Connection is distinct, we can treat each connection as a separate lock domain. that should allow rgw to scale much better with connection count

the tradeoff is that lsquic managed the set of open connections and their lookups by connection id, whereas quiche leaves that part up to the application. i _believe_ we can do this part with no locking at all

in the current prototype, i did add a mutex to the connection set in `Listener::on_packet()`, but as long as the single `Listener::receive_packets()` coroutine is the only thing that accesses that connection set, it doesn't actually need to lock

cbodley commented:
> with the rest of the code we are using the spawn library with the optiobal_yield. i thought that the plan would be to refactor the frontends with c++20 coroutines in a later stage?

this will still use the spawn library to create the stackful coroutine that calls `process_request()`, but i haven't quite gotten that far. the frontend can still use c++20 coroutines internally for its implementation, they're just not exposed to the rest of rgw

mattbenjamin commented:
> > what were the reasons to prefer quiche over lsquic ?
> 
> for me, the main differentiator was related to locking. with lsquic

I feel we should exercise some caution about incorporating a rust library, but as a prototype, it seems motivated.  I don't mind supporting it if it works well--and rust is so au courant.

Matt

cbodley commented:
it's alive and working, but the bulk data transfers are pretty slow. testing PUT and GET of a 128mb object with http3:
```
~/ceph/build $ time SSL_CERT_FILE=cert.crt ~/curl/build/src/curl --http3 --aws-sigv4 aws:amz:us-east-1:s3 -u '0555b35654ad1656d804:h7GhxuBLTrlhVUyxSPUKUV8r/2EI4ngqJxD7iBdBYLhwluN30JaT3Q==' https://localhost:8000/testbucket/128m.iso -T 128m.iso

real    0m2.123s
user    0m0.160s
sys     0m1.615s

~/ceph/build $ time SSL_CERT_FILE=cert.crt ~/curl/build/src/curl --http3 --aws-sigv4 aws:amz:us-east-1:s3 -u '0555b35654ad1656d804:h7GhxuBLTrlhVUyxSPUKUV8r/2EI4ngqJxD7iBdBYLhwluN30JaT3Q==' https://localhost:8000/testbucket/128m.iso --output 128m.out
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  128M  100  128M    0     0  13.1M      0  0:00:09  0:00:09 --:--:-- 13.3M

real    0m9.875s
user    0m0.101s
sys     0m2.976s
```
the same requests with tcp for comparison:
```
~/ceph/build $ time ~/curl/build/src/curl --aws-sigv4 aws:amz:us-east-1:s3 -u '0555b35654ad1656d804:h7GhxuBLTrlhVUyxSPUKUV8r/2EI4ngqJxD7iBdBYLhwluN30JaT3Q==' http://localhost:8000/testbucket/128m.iso -T 128m.iso

real    0m0.775s
user    0m0.000s
sys     0m0.053s

~/ceph/build $ time ~/curl/build/src/curl --aws-sigv4 aws:amz:us-east-1:s3 -u '0555b35654ad1656d804:h7GhxuBLTrlhVUyxSPUKUV8r/2EI4ngqJxD7iBdBYLhwluN30JaT3Q==' http://localhost:8000/testbucket/128m.iso --output 128m.out
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  128M  100  128M    0     0   519M      0 --:--:-- --:--:-- --:--:--  520M

real    0m0.388s
user    0m0.003s
sys     0m0.242s
```
the addition of `recvmmsg()` helped a lot with the PUTs, but most of the overhead seems to come from the timer in StreamIO which i use for streaming data between `Connection` and `ClientIO`. we're triggering a wakeup on that timer for every packet we receive, instead of once when the entire 4mb i/o completes

mattbenjamin commented:
@cbodley I assume that timer could eventually be eliminated?

cbodley commented:
> @cbodley I assume that timer could eventually be eliminated?

yeah, we just need *some* way to wake up ClientIO once their chunk of i/o completes. as a first draft, i used a timer to act like a condition variable here, but it's being notified far too often. once we convert this to a one-time notification, we can use a different primitive like `ceph::async::Completion` to avoid the need for file descriptors and system calls associated with eventfd

cbodley commented:
previous comparisons were against the beast frontend without ssl. new comparisons use beast with ssl. latencies (in seconds) are taken from the rgw log and averaged over 3 samples. single-osd vstart cluster on nvme

| frontend  | 4K PUT | 128M PUT | 1G PUT | 4K GET | 128M GET | 1G GET |
|-----------|--------|----------|--------|--------|----------|--------|
| quiche    |  0.003 |    0.828 |  6.499 |  0.001 |    0.828 |  1.963 |
| beast+ssl |  0.003 |    0.502 |  3.791 |  0.001 |    0.502 |  1.837 |

cbodley commented:
> - [x] split quiche dependency into a shared library and dlopen it to isolate the boringssl linkage

i managed to split out the boringssl+quiche bits into a shared library, and wrote cmake scripts to automate the cloning/building of boringssl, quiche, and libcurl. it works, but there's a big performance regression for PUTs that i haven't tracked down yet. once i fix that, i'll update the PR. for now, those changes are staged in https://github.com/cbodley/ceph/tree/wip-rgw-h3-external/src/rgw/h3

mkogan1 commented on src/rgw/h3/README.md:
```diff
@@ -0,0 +1,56 @@
+
+### Build Ceph with all of the above
+
+	~/ceph/build $ cmake -GNinja -DCMAKE_BUILD_TYPE=Debug -DWITH_MGR=OFF -DWITH_RADOSGW_KMIP=OFF -DWITH_RADOSGW_DBSTORE=OFF -DWITH_RADOSGW_QUICHE=ON -DOPENSSL_USE_STATIC_LIBS=ON -DOPENSSL_ROOT_DIR=$HOME/boringssl/build -DOPENSSL_INCLUDE_DIR=$HOME/boringssl/src/include -DQUICHE_LIBRARY=$HOME/quiche/target/release/libquiche.a -DQUICHE_INCLUDE_DIR=$HOME/quiche/quiche/include -DCURL_LIBRARY=$HOME/curl/b ...
```
```suggestion
	~/ceph/build $ cmake -GNinja -DCMAKE_BUILD_TYPE=Debug -DWITH_MGR=OFF -DWITH_RADOSGW_KMIP=OFF -DWITH_RADOSGW_DBSTORE=OFF -DWITH_RADOSGW_QUICHE=ON -DOPENSSL_USE_STATIC_LIBS=ON -DOPENSSL_ROOT_DIR=$HOME/boringssl/build -DOPENSSL_INCLUDE_DIR=$HOME/boringssl/src/include -DQUICHE_LIBRARY=$HOME/quiche/target/release/libquiche.a -DQUICHE_INCLUDE_DIR=$HOME/quiche/quiche/include -DCURL_LIBRARY=$HOME/curl/build/lib/libcurl.so -DCURL_INCLUDE_DIR=$HOME/curl/include ..
```

cbodley commented:
i didn't manage to resolve the performance regression, but i was able to split up the shared-library commits and isolate it to the switch from static linkage to dlopen. GET performance is similar, but the 128M PUT latency more than tripled. in both cases, `perf record` shows the majority of time spent in the kernel on msr calls

before commit `cmake/rgw: build h3 as an external project and dlopen it`:
```
Samples: 23K of event 'cycles', Event count (approx.): 15179335915
Overhead  Command          Shared Object         Symbol
  52.23%  radosgw          [kernel.kallsyms]     [k] native_write_msr
  30.67%  radosgw          [kernel.kallsyms]     [k] native_read_msr
   2.63%  msgr-worker-2    [kernel.kallsyms]     [k] native_write_msr
```
after:
```
Samples: 27K of event 'cycles', Event count (approx.): 23221356678
Overhead  Command          Shared Object              Symbol
  45.34%  radosgw          [kernel.kallsyms]          [k] native_write_msr
  20.95%  radosgw          [kernel.kallsyms]          [k] native_read_msr
  10.54%  radosgw          librgw_h3_quiche.so.1.0.0  [.] boost::container::uninitialized_default_init_alloc_n<boost::container::dtl::static_storage_allocator
   6.61%  radosgw          librgw_h3_quiche.so.1.0.0  [.] operator new
```

mkogan1 commented on cmake/modules/BuildQuiche.cmake:
```diff
@@ -0,0 +1,42 @@
+
+  # build the library with cargo, but invoke it with cmake -E env
+  # to pass in the boringssl path
+  set(cargo_command ${CMAKE_COMMAND} -E env QUICHE_BSSL_PATH=${BORINGSSL_SOURCE_DIR} -- cargo build)
```
```suggestion
  set(cargo_command ${CMAKE_COMMAND} -E env QUICHE_BSSL_PATH=${BORINGSSL_SOURCE_DIR} cargo build)
```
resolves a compile error:
```
[89/950] Performing build step for 'quiche_ext'
FAILED: src/rgw/quiche_ext-prefix/src/quiche_ext-stamp/quiche_ext-build src/rgw/quiche/target/release/libquiche.a src/rgw/quiche/target/release/quiche.pc
cd /mnt/nvme0n1p1/src-git/QUIC/ceph--up--main/build/src/rgw/quiche && /usr/bin/cmake -E env QUICHE_BSSL_PATH=/mnt/nvme0n1p1/src-git/QUIC/ceph--up--main/buil
d/src/rgw/boringssl/src -- cargo build --features ffi,pkg-config-meta,qlog --release && /usr/bin/cmake -E touch /mnt/nvme0n1p1/src-git/QUIC/ceph--up--main/b
uild/src/rgw/quiche_ext-prefix/src/quiche_ext-stamp/quiche_ext-build
cmake -E env: unknown option '--'
```

on 
```
cat /etc/redhat-release
CentOS Stream release 8
```
cmake version:
```
dnf info cmake
Installed Packages
Name         : cmake
Version      : 3.20.2
Release      : 5.el8
Architecture : x86_64
Size         : 31 M
Source       : cmake-3.20.2-5.el8.src.rpm
Repository   : @System
From repo    : appstream
...
```

cbodley replied:
ah thanks. the docs at https://cmake.org/cmake/help/latest/manual/cmake.1.html#cmdoption-cmake-e-arg-env say the double-dash is optional, and `New in version 3.24`. i'll remove those

mkogan1 commented on cmake/modules/BuildCURL.cmake:
```diff
@@ -0,0 +1,49 @@
+  set(curl_BYPRODUCTS ${curl_LIBRARY} ${curl_BINARY})
+
+  # resolve undefined reference to `fmod'
+  set(make_command ${CMAKE_COMMAND} -E env LDFLAGS="-lm" -- ${CMAKE_COMMAND})
```
```suggestion
  set(make_command ${CMAKE_COMMAND} -E env LDFLAGS="-lm" ${CMAKE_COMMAND})
```
resolves compile error:
```
[42/675] Performing configure step for 'curl_ext'
FAILED: src/rgw/curl_ext-prefix/src/curl_ext-stamp/curl_ext-configure
cd /mnt/nvme0n1p1/src-git/QUIC/ceph--up--main/build/src/rgw/curl/bin && /usr/bin/cmake -E env LDFLAGS=\"-lm\" -- /usr/bin/cmake -DHTTP_ONLY=ON -DOPENSSL_USE_STATIC_LIBS=ON -DOPENSSL_ROOT_DIR=/mnt/nvme0n1p1/src-git/QUIC/ceph--up--main/build/src/rgw/boringssl/src/build -DOPENSSL_INCLUDE_DIR=/mnt/nvme0n1p1/src-git/QUIC/ceph--up--main/build/src/rgw/boringssl/src/src/include -DUSE_QUICHE=ON -DQUICHE_LIBRARY=/mnt/nvme0n1p1/src-git/QUIC/ceph--up--main/build/src/rgw/quiche/target/release/libquiche.a -DQUICHE_INCLUDE_DIR=/mnt/nvme0n1p1/src-git/QUIC/ceph--up--main/build/src/rgw/quiche/quiche/include -DCMAKE_C_COMPILER=/opt/rh/gcc-toolset-12/root/usr/bin/gcc -DCMAKE_AR=/opt/rh/gcc-toolset-12/root/usr/bin/ar -DCMAKE_BUILD_TYPE=Release -DCMAKE_RUNTIME_OUTPUT_DIRECTORY=/mnt/nvme0n1p1/src-git/QUIC/ceph--up--main/build/bin -GNinja /mnt/nvme0n1p1/src-git/QUIC/ceph--up--main/build/src/rgw/curl/src && /usr/bin/cmake -E touch /mnt/nvme0n1p1/src-git/QUIC/ceph--up--main/build/src/rgw/curl_ext-prefix/src/curl_ext-stamp/curl_ext-configure
cmake -E env: unknown option '--'
```

mkogan1 commented:
> i didn't manage to resolve the performance regression, but i was able to split up the shared-library commits and isolate it to the switch from static linkage to dlopen. GET performance is similar, but the 128M PUT latency more than tripled. in both cases, `perf record` shows the majority of time spent in the kernel on msr calls
> 
> before commit `cmake/rgw: build h3 as an external project and dlopen it`:
> 
> ```
> Samples: 23K of event 'cycles', Event count (approx.): 15179335915
> Overhead  Command          Shared Object         Symbol
>   52.23%  radosgw          [kernel.kallsyms]     [k] native_write_msr
>   30.67%  radosgw          [kernel.kallsyms]     [k] native_read_msr
>    2.63%  msgr-worker-2    [kernel.kallsyms]     [k] native_write_msr
> ```
> 
> after:
> 
> ```
> Samples: 27K of event 'cycles', Event count (approx.): 23221356678
> Overhead  Command          Shared Object              Symbol
>   45.34%  radosgw          [kernel.kallsyms]          [k] native_write_msr
>   20.95%  radosgw          [kernel.kallsyms]          [k] native_read_msr
>   10.54%  radosgw          librgw_h3_quiche.so.1.0.0  [.] boost::container::uninitialized_default_init_alloc_n<boost::container::dtl::static_storage_allocator
>    6.61%  radosgw          librgw_h3_quiche.so.1.0.0  [.] operator new
> ```

will check perf record on my system with various callstack recording mechanisms enabled and see if can pinpoint the source, (alternatively can try *trace utils & ebpf)

if had to guess, think it is probably modifying the same MSR resisters as `sudo cpupower idle-set ...` / `sudo cpupower idle-info` to control the latency

mkogan1 commented on src/rgw/h3/CMakeLists.txt:
```diff
@@ -0,0 +1,21 @@
+
+find_package(OpenSSL REQUIRED)
+find_package(Quiche 0.16.0 REQUIRED)
+find_package(Boost 1.81 REQUIRED COMPONENTS system)
```
to satisfy this, had to change:
![image](https://user-images.githubusercontent.com/31659604/220617286-d5f1661b-156f-4b2b-831e-741377427420.png)
but then there are link errors:
```
/usr/local/libexec/gold/ld: error: lib/libcommon.a(ConfUtils.cc.o): multiple definition of 'boost::phoenix::placeholders::uarg10'
/usr/local/libexec/gold/ld: lib/libosd.a(OSDCap.cc.o): previous definition here
/usr/local/libexec/gold/ld: error: lib/libcommon.a(ConfUtils.cc.o): multiple definition of 'boost::phoenix::placeholders::uarg9'
/usr/local/libexec/gold/ld: lib/libosd.a(OSDCap.cc.o): previous definition here
/usr/local/libexec/gold/ld: error: lib/libcommon.a(ConfUtils.cc.o): multiple definition of 'boost::phoenix::placeholders::uarg8'
/usr/local/libexec/gold/ld: lib/libosd.a(OSDCap.cc.o): previous definition here
/usr/local/libexec/gold/ld: error: lib/libcommon.a(ConfUtils.cc.o): multiple definition of 'boost::phoenix::placeholders::uarg7'
/usr/local/libexec/gold/ld: lib/libosd.a(OSDCap.cc.o): previous definition here
/usr/local/libexec/gold/ld: error: lib/libcommon.a(ConfUtils.cc.o): multiple definition of 'boost::phoenix::placeholders::uarg6'
/usr/local/libexec/gold/ld: lib/libosd.a(OSDCap.cc.o): previous definition here
/usr/local/libexec/gold/ld: error: lib/libcommon.a(ConfUtils.cc.o): multiple definition of 'boost::phoenix::placeholders::uarg5'
/usr/local/libexec/gold/ld: lib/libosd.a(OSDCap.cc.o): previous definition here
/usr/local/libexec/gold/ld: error: lib/libcommon.a(ConfUtils.cc.o): multiple definition of 'boost::phoenix::placeholders::uarg4'
/usr/local/libexec/gold/ld: lib/libosd.a(OSDCap.cc.o): previous definition here
/usr/local/libexec/gold/ld: error: lib/libcommon.a(ConfUtils.cc.o): multiple definition of 'boost::phoenix::placeholders::uarg3'
/usr/local/libexec/gold/ld: lib/libosd.a(OSDCap.cc.o): previous definition here
/usr/local/libexec/gold/ld: error: lib/libcommon.a(ConfUtils.cc.o): multiple definition of 'boost::phoenix::placeholders::uarg2'
/usr/local/libexec/gold/ld: lib/libosd.a(OSDCap.cc.o): previous definition here
/usr/local/libexec/gold/ld: error: lib/libcommon.a(ConfUtils.cc.o): multiple definition of 'boost::phoenix::placeholders::uarg1'
/usr/local/libexec/gold/ld: lib/libosd.a(OSDCap.cc.o): previous definition here
/usr/local/libexec/gold/ld: error: lib/libcommon.a(MonCap.cc.o): multiple definition of 'boost::phoenix::placeholders::uarg10'
/usr/local/libexec/gold/ld: lib/libosd.a(OSDCap.cc.o): previous definition here
/usr/local/libexec/gold/ld: error: lib/libcommon.a(MonCap.cc.o): multiple definition of 'boost::phoenix::placeholders::uarg9'
/usr/local/libexec/gold/ld: lib/libosd.a(OSDCap.cc.o): previous definition here
/usr/local/libexec/gold/ld: error: lib/libcommon.a(MonCap.cc.o): multiple definition of 'boost::phoenix::placeholders::uarg8'
/usr/local/libexec/gold/ld: lib/libosd.a(OSDCap.cc.o): previous definition here
/usr/local/libexec/gold/ld: error: lib/libcommon.a(MonCap.cc.o): multiple definition of 'boost::phoenix::placeholders::uarg7'
/usr/local/libexec/gold/ld: lib/libosd.a(OSDCap.cc.o): previous definition here
/usr/local/libexec/gold/ld: error: lib/libcommon.a(MonCap.cc.o): multiple definition of 'boost::phoenix::placeholders::uarg6'
/usr/local/libexec/gold/ld: lib/libosd.a(OSDCap.cc.o): previous definition here
/usr/local/libexec/gold/ld: error: lib/libcommon.a(MonCap.cc.o): multiple definition of 'boost::phoenix::placeholders::uarg5'
/usr/local/libexec/gold/ld: lib/libosd.a(OSDCap.cc.o): previous definition here
/usr/local/libexec/gold/ld: error: lib/libcommon.a(MonCap.cc.o): multiple definition of 'boost::phoenix::placeholders::uarg4'
/usr/local/libexec/gold/ld: lib/libosd.a(OSDCap.cc.o): previous definition here
/usr/local/libexec/gold/ld: error: lib/libcommon.a(MonCap.cc.o): multiple definition of 'boost::phoenix::placeholders::uarg3'
/usr/local/libexec/gold/ld: lib/libosd.a(OSDCap.cc.o): previous definition here
/usr/local/libexec/gold/ld: error: lib/libcommon.a(MonCap.cc.o): multiple definition of 'boost::phoenix::placeholders::uarg2'
/usr/local/libexec/gold/ld: lib/libosd.a(OSDCap.cc.o): previous definition here
/usr/local/libexec/gold/ld: error: lib/libcommon.a(MonCap.cc.o): multiple definition of 'boost::phoenix::placeholders::uarg1'
/usr/local/libexec/gold/ld: lib/libosd.a(OSDCap.cc.o): previous definition here
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.

```

cbodley replied:
oops, i had been experimenting with this on top of the wip-coro-after-reef feature branch where we updated to 1.81. i switched it back to 1.79

cbodley commented:
@mkogan1 exposed `ack_delay_exponent` as an rgw_frontends option (ie `quiche port=80 ack_delay_exponent=30`) to experiment with the "Reducing acknowledgement frequency" optimization mentioned in https://www.fastly.com/blog/measuring-quic-vs-tcp-computational-efficiency

cbodley committed:
rgw/h3: add h3_zero example

Signed-off-by: Casey Bodley <cbodley@redhat.com>

cbodley commented:
@mkogan1 i added a new commit with an example program `h3_zero`. that's a super-simple GET/PUT/HEAD api server that i think could be useful for performance measurements without any overhead from ceph or rgw; it doesn't store any data, and GET requests just return zeroes. let me know if you have any ideas to make it more useful

cbodley committed:
DNM: enable WITH_RADOSGW_QUICHE by default for CI

cbodley committed:
rgw/beast: add frontend option for Alt-Svc header

this allows http/1.1 clients to discover the presence of an http/3 endpoint

Signed-off-by: Casey Bodley <cbodley@redhat.com>

cbodley committed:
rgw/h3: enable key logging if SSLKEYLOGFILE is in env

Signed-off-by: Casey Bodley <cbodley@redhat.com>

cbodley committed:
rgw/h3: frontend exposes and documents all options

the cert= and key= options were renamed to match beast's
ssl_certificate= and ssl_private_key=

Signed-off-by: Casey Bodley <cbodley@redhat.com>

mkogan1 commented:
> @mkogan1 i added a new commit with an example program `h3_zero`. that's a super-simple GET/PUT/HEAD api server that i think could be useful for performance measurements without any overhead from ceph or rgw; it doesn't store any data, and GET requests just return zeroes. let me know if you have any ideas to make it more useful

@cbodley good update, with `h3_zero` the 1:1 ratio between data and ack packets does not occur
![image](https://user-images.githubusercontent.com/31659604/230941041-b9cf0043-d7c5-4a71-a427-4b6cdd11350a.png)
working on the diff diag

mkogan1 commented on src/rgw/h3/src/connection.cc:
```diff
@@ -0,0 +1,555 @@
+
+  if (send_at > pacing_clock::zero()) {
+    const auto now = pacing_clock::now();
+    if (send_at > now) {
```
in testing, the condition code flow wasn't executing - `send_at` was never larger than `now`
```suggestion
    if (now > send_at) {
```

mkogan1 commented on src/rgw/h3/src/connection.cc:
```diff
@@ -0,0 +1,555 @@
+  if (send_at > pacing_clock::zero()) {
+    const auto now = pacing_clock::now();
+    if (send_at > now) {
+      const auto delay = send_at - now;
```
```suggestion
      const auto delay = now - send_at;
```

cbodley commented:
changes:
* rebased over asio::spawn() changes
* bumped quiche version from 0.16.0 to 0.20.0
* reenabled pacing delays which were broken on 0.16.0

cbodley committed:
rgw/zero: fix error handling for /zero api

dump_body() and recv_body() both catch exceptions and return negative
error codes. handle negative values as errors instead of trying to catch
them

errors here led to later crashes when trying to use a null s->formatter
to format the error response. s->formatter is now allowed on init()

Signed-off-by: Casey Bodley <cbodley@redhat.com>

cbodley committed:
rgw/rest: also catch i/o exceptions as boost system_error

rgw frontends throw rgw::io::Exception = std::system_error for io
errors, but other internal exceptions from boost::system::system_error
may also escape from ClientIO. instead of trying to catch and rethrow as
rgw::io::Exception, catch both types in the rgw_rest.cc helpers

Signed-off-by: Casey Bodley <cbodley@redhat.com>

cbodley commented:
performance looks fine against the h3_zero example server. it's also fine for GET requests to rgw's /zero api, but PUTs are extremely slow. wireshark shows tons of tiny packets being exchanged there, ie separate ACKs of individual packets

cbodley committed:
rgw/h3: support concurrent stream read and write

StreamIO tracks read and write state separately so that it can handle a
read and a write at the same time

Signed-off-by: Casey Bodley <cbodley@redhat.com>

cbodley committed:
rgw/h3: Listener dispatches packets in batches

each time the Listener calls recvmmsg() to read an array of packets, it
groups those packets by their associated Connection. this allows it to
deliver all of its packets to the Connection in a single call to
async_handle_packets()

this avoids the need for multiple round trips between the Listener's
executor and the Connection's. it also allows the Connection to process
all of the packets before calling writer_wake() to send response packets
or acks

Signed-off-by: Casey Bodley <cbodley@redhat.com>

cbodley committed:
rgw/h3: use ip::udp::socket

Signed-off-by: Casey Bodley <cbodley@redhat.com>

RaminNietzsche commented:
@cbodley — rebased `wip-rgw-h3` onto current `main` and resolved conflicts on a fork branch:

**Branch:** `RaminNietzsche/ceph:wip-rgw-h3` (10 commits; DNM CI commit dropped)

**Rebase / conflict resolution:**
- Rebased onto `ceph/main` (2026-06-22)
- **`rgw_appmain.cc`:** use `context_pool_holder.get()` (matches current beast/kms path); fixed `#ifdef HAVE_QUICHE` / `quiche` frontend registration (was closing `rgw-nfs` block)
- **`rgw_asio_{client,frontend}.cc`:** merged Alt-Svc headers with main's `keepalive` / `expect100continue` / `optional_yield` beast async path
- **`doc/radosgw/frontends.rst`:** kept `:ref:` style + Quiche section + Alt-Svc/TLS links
- **Skipped** `DNM: enable WITH_RADOSGW_QUICHE by default for CI` — `WITH_RADOSGW_QUICHE` stays **OFF**

**Suggested next steps:**
1. `git fetch https://github.com/RaminNietzsche/ceph.git wip-rgw-h3`
2. Reset your branch or cherry-pick the series
3. `jenkins test make check` (with `-DWITH_RADOSGW_QUICHE=ON`)
4. Functional test with `curl --http3` per `src/rgw/h3/README.md`

Experimental HTTP/3 frontend — happy to help with further review items (unit/qa tests, i/o optimizations from PR TODO).

mkogan1 commented:
@RaminNietzsche 

after `❯ git cherry-pick -ex 70d9eef9b383c8f4475051d8b1ce124dd20a617d^..fdc41e70618aea80bc630724197a898b3333425b`

there is a build error
```
FAILED: src/rgw/CMakeFiles/rgw_a.dir/rgw_asio_client.cc.o 
/usr/bin/ccache /opt/rh/gcc-toolset-15/root/usr/bin/g++ -DBOOST_ASIO_DISABLE_THREAD_KEYWORD_EXTENSION -DBOOST_ASIO_NO_TS_EXECUTORS -DBOOST_USE_VALGRIND -DCLS_CLIENT_HIDE_IOCTX -DHAVE_CONFIG_H -DXXH_NO_INLINE_HINTS=1 -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -D_REENTRANT -D_THREAD_SAFE -D__CEPH__ -D__STDC_FORMAT_MACROS -D__linux__ -I/mnt/sdb/src-git/ceph--up/build/src/include -I/mnt/sdb/src-git/ceph--up/src -I/mnt/sdb/src-git/ceph--up/src/dmclock/support/src -I/mnt/sdb/src-git/ceph--up/src/rgw -I/mnt/sdb/src-git/ceph--up/src/rgw/driver/rados -I/mnt/sdb/src-git/ceph--up/src/libkmip -I/mnt/sdb/src-git/ceph--up/src/rgw/services -I/mnt/sdb/src-git/ceph--up/src/BLAKE3/c -I/mnt/sdb/src-git/ceph--up/src/dmclock/src -isystem /mnt/sdb/src-git/ceph--up/build/boost/include -isystem /mnt/sdb/src-git/ceph--up/build/include -isystem /mnt/sdb/src-git/ceph--up/src/xxHash -isystem /mnt/sdb/src-git/ceph--up/src/fmt/include -isystem /mnt/sdb/src-git/ceph--up/src/s3select/rapidjson/include -O2 -g -DNDEBUG -std=c++23 -fPIC -fdiagnostics-color=always -fno-builtin-malloc -fno-builtin-calloc -fno-builtin-realloc -fno-builtin-free -Wall -fno-strict-aliasing -fsigned-char -Wtype-limits -Wignored-qualifiers -Wpointer-arith -Werror=format-security -Winit-self -Wno-unknown-pragmas -Wnon-virtual-dtor -Wno-ignored-qualifiers -ftemplate-depth-1024 -DBOOST_ALLOW_DEPRECATED_HEADERS -Wpessimizing-move -Wredundant-move -Wstrict-null-sentinel -Woverloaded-virtual -fstack-protector-strong -fdiagnostics-color=always -Werror=vla -MD -MT src/rgw/CMakeFiles/rgw_a.dir/rgw_asio_client.cc.o -MF src/rgw/CMakeFiles/rgw_a.dir/rgw_asio_client.cc.o.d -o src/rgw/CMakeFiles/rgw_a.dir/rgw_asio_client.cc.o -c /mnt/sdb/src-git/ceph--up/src/rgw/rgw_asio_client.cc
/mnt/sdb/src-git/ceph--up/src/rgw/rgw_asio_client.cc: In member function ‘virtual size_t rgw::asio::ClientIO::complete_header()’:
/mnt/sdb/src-git/ceph--up/src/rgw/rgw_asio_client.cc:162:16: error: expression cannot be used as a function
  162 |   if (keepalive()) {
      |       ~~~~~~~~~^~
```
myb it should be `keep_alive()` ?

RaminNietzsche commented:
@mkogan1 Thanks for catching this — you're right, it should be `keep_alive()`.

That was my mistake in the Alt-Svc / beast merge: I called the `keepalive` member as if it were a function instead of using the `keep_alive()` accessor from `rgw_asio_client.h`.

Fixed in `RaminNietzsche/ceph:wip-rgw-h3` (commit `78ad5c93f92`). I rebuilt locally and `make` completed cleanly, including:

```
[591/591 (100%) ] Linking CXX executable bin/unittest_rgw_cors
```

Sorry for the noise, and thanks again for testing the cherry-pick series.
